### PR TITLE
Persist delivery config and associated things transactionally

### PR DESCRIPTION
**Pull request by robfletcher**
… at least when using SQL.

### CODE REVIEW

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt [hunk 1]
```diff
@@ -10,39 +10,40 @@ import com.netflix.spinnaker.keel.persistence.metamodel.Tables.DELIVERY_ARTIFACT
 import com.netflix.spinnaker.keel.persistence.metamodel.Tables.DELIVERY_ARTIFACT_VERSION
 import org.jooq.DSLContext
 import org.slf4j.LoggerFactory
+import org.springframework.transaction.annotation.Propagation.REQUIRED
+import org.springframework.transaction.annotation.Transactional
 
-class SqlArtifactRepository(
+open class SqlArtifactRepository(
```
Comment (robfletcher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Still trying to figure out why I had to do this when the `kotlin-spring` plugin is applied and it works in `ResourcePersister`.

Comment (robfletcher) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt]: Seems to be something to do with the fact the annotations are on methods defined in the interface.

File: keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt [hunk 2]
```diff
@@ -17,27 +23,53 @@ import com.netflix.spinnaker.keel.plugin.supporting
...
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val artifactRepository: ArtifactRepository,
   private val resourceRepository: ResourceRepository,
   private val handlers: List<ResolvableResourceHandler<*, *>>,
   private val clock: Clock,
   private val publisher: ApplicationEventPublisher
 ) {
-  fun upsert(resource: SubmittedResource<Any>): Resource<out Any> =
+  @Transactional(propagation = REQUIRED)
+  fun upsert(deliveryConfig: SubmittedDeliveryConfig): DeliveryConfig =
```
Comment (robfletcher) [inline keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt]: This is the key method that needs to be transactional.

File: keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/tx/DeliveryConfigTransactionTests.kt [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  webEnvironment = MOCK,
+  properties = [
+    "sql.enabled=true",
+    "sql.connection-pools.default.jdbc-url=jdbc:tc:mysql:5.7.22://somehostname:someport/databasename",
+    "sql.migration.jdbc-url=jdbc:tc:mysql:5.7.22://somehostname:someport/databasename",
+    "spring.datasource.driver-class-name=org.testcontainers.jdbc.ContainerDatabaseDriver",
+    "eureka.default-to-up=false"
+  ]
+)
+internal class DeliveryConfigTransactionTests : JUnit5Minutests {
```
Comment (robfletcher) [inline keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/tx/DeliveryConfigTransactionTests.kt]: Here are a bunch of tests to make sure delivery config is persisted all or nothing

File: keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/tx/DeliveryConfigTransactionTests.kt [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+  @SpykBean
+  lateinit var artifactRepository: ArtifactRepository
+
+  @SpykBean
+  lateinit var resourceRepository: ResourceRepository
+
+  @SpykBean
```
Comment (robfletcher) [inline keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/tx/DeliveryConfigTransactionTests.kt]: First time I've ever really used spies. I want the persistence to actually work, but also to be able to force exceptions at certain places when I want to

File: keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt (hunk 2 above)
Comment (asher) [reply keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt]: `SqlDeliveryConfigRepository.store()` doesn't appear to be transaction though, so I feel like I'm missing something?

Comment (robfletcher) [reply keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt]: It doesn't need to be to operate in the transaction that this method creates.

Comment (asher) [reply keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt]: I missed that spring-tx is managing transaction boundaries here 👍

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt [hunk 5]
```diff
@@ -29,43 +29,35 @@ class SqlDeliveryConfigRepository(
...
-            .set(ENVIRONMENT.DELIVERY_CONFIG_UID, uid)
-            .set(ENVIRONMENT.NAME, environment.name)
+      }
+      environments.forEach { environment ->
+        val environmentUID = randomUID().toString()
+        jooq.insertInto(ENVIRONMENT)
+          .set(ENVIRONMENT.UID, environmentUID)
+          .set(ENVIRONMENT.DELIVERY_CONFIG_UID, uid)
+          .set(ENVIRONMENT.NAME, environment.name)
+          .onDuplicateKeyIgnore()
```
Comment (asher) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt]: Not sure about environments always getting a new uid here. This store method is called for delivery config upsert, so it seems like the environment table is likely to get lots of "duplicate" rows, where `delivery_config_uid` and `name` are the same, but with a new `uid` per config upsert.

**Comment by robfletcher** [comment]
Good spot, yeah. Will fix.

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/DeliveryConfigRepositoryTests.kt [hunk 6]
```diff
@@ -132,30 +143,48 @@ abstract class DeliveryConfigRepositoryTests<T : DeliveryConfigRepository, R : R
...
+          storeArtifacts()
+          storeResources()
+          store()
+          store()
+        }
+
+        test("the environments are not duplicated") {
+          getByName()
+            .succeeded()
+            .get { environments }.hasSize(deliveryConfig.environments.size)
```
Comment (robfletcher) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/DeliveryConfigRepositoryTests.kt]: Test to ensure environments are not re-created

**Pull request merged**